gossip/client/hubclient_test.go

```diff
@@ -0,0 +1,627 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client_test
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/trillian-examples/gossip/api"
+	"github.com/google/trillian-examples/gossip/client"
+	"github.com/google/trillian/testonly"
+
+	tcrypto "github.com/google/trillian/crypto"
+)
+
+// serveAt returns a test HTTP server that returns a canned response status/body for a given path.
+func serveAt(t *testing.T, path string, status int, body string) *httptest.Server {
+	t.Helper()
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != api.PathPrefix+path {
+			t.Fatalf("Incorrect URL path: %s", r.URL.Path)
+		}
+		if status != 0 {
+			w.WriteHeader(status)
+		}
+		fmt.Fprintf(w, body)
+	}))
+}
+
+// generateKeys creates a new ECDSA private key and returns it as a tcrypto.Signer,
+// together with the DER for the associated public key in PKIX format.
+func generateKeys(t *testing.T) (*tcrypto.Signer, []byte) {
+	t.Helper()
+	ecdsaKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		t.Fatalf("Failed to generate ECDSA key: %v", err)
+	}
+	pubKeyDER, err := x509.MarshalPKIXPublicKey(ecdsaKey.Public())
+	if err != nil {
+		t.Fatalf("Failed to marshal public key to DER: %v", err)
+	}
+	signer := tcrypto.NewSigner(6962, ecdsaKey, crypto.SHA256)
+	return signer, pubKeyDER
+}
+
+// signData returns data, signature-over-data as a 2 element slice of base-64 encoded strings.
+func signData(t *testing.T, signer *tcrypto.Signer, data []byte) []interface{} {
```

Comment:
I don't understand why the return type is `[]interface{}`. If there's a good reason for it, it should probably go in the function's documentation.

Reply:
Yeah, this could have done with a note -- returning as a slice of `interface{}` means that the results can be passed to variadic functions like `Sprintf` (e.g. `fmt.Sprintf("%s %s", signData(t, s, x)...)` )

Reply:
Fixed in #36